image/ani.cpp

```diff
@@ -0,0 +1,304 @@
+		warning("AniDecoder::parseRIFFContainer: RIFF container is too small");
+		return false;
+	}
+
+	byte containerTypeID[4];
+	if (chunkStream.read(containerTypeID, 4) != 4) {
+		warning("AniDecoder::parseRIFFContainer: Failed to read RIFF container type");
+		return false;
+	}
+
+	RIFFContainerDef containerDef;
+	containerDef.id = READ_BE_UINT32(containerTypeID);
+	containerDef.size = chunkDef.size - 4;
+
+	Common::SeekableSubReadStream substream(&chunkStream, 4, chunkDef.size);
+	return callback(containerDef, substream);
+}
+
+bool AniDecoder::parseTopLevelChunk(const RIFFChunkDef &chunk, Common::SeekableReadStream &stream) {
+	if (chunk.id != MKTAG('R', 'I', 'F', 'F')) {
+		warning("AniDecoder::parseTopLevelChunk: Top-level chunk isn't RIFF");
+		return false;
+	}
+
+	return parseRIFFContainer(stream, chunk, Common::Functor2Mem<const RIFFContainerDef &, Common::SeekableReadStream &, bool, AniDecoder>(this, &AniDecoder::parseTopLevelContainer));
+}
+
+bool AniDecoder::parseTopLevelContainer(const RIFFContainerDef &container, Common::SeekableReadStream &stream) {
+	if (container.id == MKTAG('A', 'C', 'O', 'N'))
+		return parseRIFFChunks(stream, Common::Functor2Mem<const RIFFChunkDef &, Common::SeekableReadStream &, bool, AniDecoder>(this, &AniDecoder::parseSecondLevelChunk));
+
+	warning("AniDecoder::parseTopLevelContainer: Top-level container isn't ACON");
+	return false;
+}
+
+bool AniDecoder::parseSecondLevelChunk(const RIFFChunkDef &chunk, Common::SeekableReadStream &stream) {
+	if (chunk.id == MKTAG('L', 'I', 'S', 'T'))
+		return parseRIFFContainer(stream, chunk, Common::Functor2Mem<const RIFFContainerDef &, Common::SeekableReadStream &, bool, AniDecoder>(this, &AniDecoder::parseListContainer));
+
+	if (chunk.id == MKTAG('a', 'n', 'i', 'h'))
+		return parseAnimHeaderChunk(chunk, stream);
+
+	if (chunk.id == MKTAG('s', 'e', 'q', ' '))
+		return parseSeqChunk(chunk, stream);
+
+	if (chunk.id == MKTAG('r', 'a', 't', 'e'))
+		return parseRateChunk(chunk, stream);
+
+	return true;
+}
+
+bool AniDecoder::parseListContainer(const RIFFContainerDef &container, Common::SeekableReadStream &stream) {
+	if (container.id == MKTAG('f', 'r', 'a', 'm'))
+		return parseRIFFChunks(stream, Common::Functor2Mem<const RIFFChunkDef &, Common::SeekableReadStream &, bool, AniDecoder>(this, &AniDecoder::parseIconChunk));
+
+	return true;
+}
+
+bool AniDecoder::parseAnimHeaderChunk(const RIFFChunkDef &chunk, Common::SeekableReadStream &stream) {
+	byte animHeader[36];
+	for (byte &b : animHeader)
+		b = 0;
+
+	uint32 amountToRead = 36;
+	if (chunk.size < amountToRead)
+		amountToRead = chunk.size;
+
+	if (amountToRead > 0 && stream.read(animHeader, amountToRead) != amountToRead) {
+		warning("AniDecoder::parseAnimHeaderChunk: Read failed");
+		return false;
+	}
+
+	uint32 structSize = READ_LE_UINT32(animHeader);
+	if (structSize < 36) {
+		for (uint i = structSize; i < 36; i++)
+			animHeader[i] = 0;
+	}
+
+	_metadata.numFrames = READ_LE_UINT32(animHeader + 4);
```

Comment:
I'd wrap it into MemoryReadStream for cleaner code. Those offsets are error-prone.